File: backend_services/app/Models/level.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Concerns\HasUuids;

class Level extends Model
{
    use HasUuids;

    protected $primaryKey = 'level_id';
    public $incrementing = false;
    protected $keyType = 'string';

    protected $fillable = [
        'level_name',
        'level_type_id',
        'level_data',
        'win_condition',
        'created_by',
        'timer'
    ];

    protected function casts(): array
    {
        return [
            'created_at' => 'datetime',
            'updated_at' => 'datetime',
        ];
    }

    // Relationship to level_type
    public function level_type()
    {
        return $this->belongsTo(level_type::class, 'level_type_id', 'level_type_id');
    }

    // Relationship to classes (many-to-many through class_levels)
    public function classes()
    {
        return $this->belongsToMany(ClassModel::class, 'class_levels', 'level_id', 'class_id')
                    ->withPivot('is_private', 'created_at', 'updated_at')
                    ->withTimestamps();
    }

    // Relationship to creator (user who created this level)
    public function creator()
    {
        return $this->belongsTo(User::class, 'created_by', 'user_id');
    }
}

File: code_play/lib/models/level.dart
class LevelModel {
  final String? levelId;
  final String? levelName;
  final String? levelTypeId;
  final String? levelTypeName;
  final String? levelData;
  final String? winCondition;
  final bool? isPrivate;
  final bool? isCreatedByMe;
  final String? status;

  LevelModel({
    this.levelId,
    this.levelName,
    this.levelTypeId,
    this.levelTypeName,
    this.levelData,
    this.winCondition,
    this.isPrivate,
    this.isCreatedByMe,
    this.status,
  });

  factory LevelModel.fromJson(Map<String, dynamic> json) {
    return LevelModel(
      levelId: json['level_id'] as String?,
      levelName: json['level_name'] as String?,
      levelTypeId: json['level_type'] != null
          ? json['level_type']['level_type_id'] as String?
          : null,
      levelTypeName: json['level_type'] != null
          ? json['level_type']['level_type_name'] as String?
          : null,
      levelData: json['level_data'] as String?,
      winCondition: json['win_condition'] as String?,
      isPrivate: json['is_private'] as bool?,
      isCreatedByMe: json['is_created_by_me'] as bool?,
      status: json['status'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'level_id': levelId,
      'level_name': levelName,
      'level_type': {
        'level_type_id': levelTypeId,
        'level_type_name': levelTypeName,
      },
      'level_data': levelData,
      'win_condition': winCondition,
    };
  }
}
